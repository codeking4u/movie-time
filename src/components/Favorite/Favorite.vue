<template>
  <div class="favorite">
    <div class="favorite__img">
      <img :src="image.medium" alt="Show image" />
    </div>
    <div class="favDetails flex flex-jc-sb">
      <p class="favDetails__name">{{ name }}</p>
      <div class="favDetails__delete flex flex-jc-c flex-ai-c">
        <button class="delete-icon" @click="removeItem(id)">X</button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import type { Favorite } from '@/types/TvShow'

export default defineComponent({
  name: 'Favorite',
  props: {
    favorite: {
      type: Object as () => Favorite,
      required: true
    }
  },
  emits: ['delete'],
  methods: {
    removeItem(index: number) {
      this.$emit('delete', index)
    }
  },
  setup(props) {
    const { id, name, image }: Favorite = props.favorite

    return {
      id,
      name,
      image
    }
  }
})
</script>

<style scoped lang="scss">
.favorite {
  display: flex;
  align-items: center;
  margin-bottom: 10px;

  &:hover {
    .delete-icon {
      visibility: visible;
    }
  }
  &__img {
    margin-right: 10px;
    img {
      width: 90px;
      height: 90px;
      object-fit: cover;
    }
  }
}

.favDetails {
  width: 100%;
  &__name {
    font-size: 1rem;
    padding-right: 2rem;
  }

  &__delete {
    margin-left: auto;
    font-size: 1rem;
    cursor: pointer;
    .delete-icon {
      visibility: hidden;
      &:hover {
        color: #8f0404;
      }
    }
  }
}
</style>
