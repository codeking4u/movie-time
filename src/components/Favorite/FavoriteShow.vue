<template>
  <router-link :to="{ name: 'ShowDetail', params: { id } }" class="favorite flex flex-ai-c">
    <div class="favorite__img">
      <img :src="image?.medium || image?.original || SHOW_DEFAULT_PIC" alt="Show image" />
    </div>
    <div class="favDetails flex flex-jc-sb">
      <p class="favDetails__name">{{ name }}</p>
      <div class="favDetails__delete flex flex-jc-c flex-ai-c">
        <button class="delete-icon" @click.prevent="() => removeItem(id)">
          <i :class="'material-icons'">clear</i>
        </button>
      </div>
    </div>
  </router-link>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import type { Favorite } from '@/types/TvShow'

export default defineComponent({
  name: 'Favorite',
  props: {
    favorite: {
      type: Object as () => Favorite,
      required: true
    }
  },
  emits: ['delete'],
  methods: {
    removeItem(index: number) {
      this.$emit('delete', index)
    }
  },
  setup(props) {
    const { id, name, image }: Favorite = props.favorite
    const SHOW_DEFAULT_PIC = import.meta.env.VITE_SHOW_DEFAULT_PIC

    return {
      id,
      name,
      image,
      SHOW_DEFAULT_PIC
    }
  }
})
</script>

<style scoped lang="scss">
.favorite {
  margin-bottom: 10px;
  max-width: 350px;
  color: #000;
  &:hover {
    background: #f4f4f4;

    .favDetails__name {
      text-decoration: underline;
    }
    .delete-icon {
      visibility: visible;
    }
  }
  &__img {
    margin-right: 10px;

    img {
      width: 90px;
      height: 90px;
      object-fit: cover;
    }
  }
}

.favDetails {
  width: 100%;
  &__name {
    font-size: 1rem;
    padding-right: 2rem;
  }

  &__delete {
    margin-left: auto;
    font-size: 1rem;
    cursor: pointer;
    .delete-icon {
      visibility: hidden;
      padding: 0px 4px;
      @include breakpoint-down(small) {
        visibility: visible;
      }

      &:hover {
        color: #8f0404;
      }
    }
  }
}
</style>
