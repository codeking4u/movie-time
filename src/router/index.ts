import { createRouter, createWebHistory } from 'vue-router'
import HomePage from '../views/HomePage.vue'
import SearchResult from '../views/SearchResult.vue'
import ShowDetail from '../views/ShowDetail.vue'

const routes = [
  {
    path: '/',
    name: 'HomePage',
    component: HomePage,
    meta: { title: 'Dashboard' }
  },
  {
    path: '/search',
    name: 'SearchResult',
    component: SearchResult,
    meta: { title: 'Search' }
  },
  {
    path: '/show/:id',
    name: 'ShowDetail',
    component: ShowDetail,
    props: true,
    meta: { title: 'Show' }
  }
]
const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: routes
})

export default router
