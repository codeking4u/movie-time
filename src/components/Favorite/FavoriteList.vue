<template>
  <div class="favorite-container section">
    <h2>Favourites</h2>
    <template v-if="allFavorites.length">
      <div class="favorite-list flex">
        <FavoriteShow
          v-for="fav in allFavorites"
          :key="fav.id"
          :favorite="fav"
          @delete="removeFavorite"
        />
      </div>
    </template>
    <template v-else>
      <p>No favourites added yet. Click on <i class="material-icons">favorite</i> icon to add.</p>
    </template>
  </div>
</template>

<script lang="ts">
import { computed } from 'vue'
import { useStore } from 'vuex'
import FavoriteShow from './FavoriteShow.vue'

export default {
  components: { FavoriteShow },
  setup() {
    const store = useStore()
    const allFavorites = computed(() => store.getters.getAllFavorites)

    const removeFavorite = (index: Number) => {
      console.log('delete')
      store.commit('removeFromFavorites', index)
    }

    return {
      allFavorites,
      removeFavorite
    }
  }
}
</script>

<style lang="scss" scoped>
.favorite-list {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
  align-items: center;
  grid-gap: 25px;
}
</style>
