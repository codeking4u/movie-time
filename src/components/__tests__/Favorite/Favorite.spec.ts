import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import Favorite from '@/components/Favorite/Favorite.vue'
import router from '@/router'

describe('Favorite.vue', () => {
  it('renders correctly', () => {
    const wrapper = mount(Favorite, {
      props: {
        favorite: {
          id: 1,
          name: 'Example Show',
          image: {
            medium: 'https://example.com/medium.jpg',
            original: 'https://example.com/original.jpg'
          }
        }
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
