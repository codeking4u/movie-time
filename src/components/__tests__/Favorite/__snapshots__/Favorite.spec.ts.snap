// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Favorite.vue > renders correctly 1`] = `
"<a data-v-73416b1a=\\"\\" href=\\"/show/1\\" class=\\"favorite flex flex-ai-c\\">
  <div data-v-73416b1a=\\"\\" class=\\"favorite__img\\"><img data-v-73416b1a=\\"\\" src=\\"https://example.com/medium.jpg\\" alt=\\"Show image\\"></div>
  <div data-v-73416b1a=\\"\\" class=\\"favDetails flex flex-jc-sb\\">
    <p data-v-73416b1a=\\"\\" class=\\"favDetails__name\\">Example Show</p>
    <div data-v-73416b1a=\\"\\" class=\\"favDetails__delete flex flex-jc-c flex-ai-c\\"><button data-v-73416b1a=\\"\\" class=\\"delete-icon\\"><i data-v-73416b1a=\\"\\" class=\\"material-icons\\">clear</i></button></div>
  </div>
</a>"
`;
